lib/travis/api/v3/queries/log.rb

```diff
@@ -1,16 +1,36 @@
 module Travis::API::V3
   class Queries::Log < Query
+    require 'net/http'
+    require 'uri'
+
     def find(job)
-      #TODO check for the log in the DB, if it's not there then fetch it from S3.
+      #check for the log in the DB
       log = Models::Log.find_by_job_id
-      #if the log exists and has not bee archived yet, then collect the log_parts and return the contents
+      #if the log exists and has not been archived yet, then collect the log_parts and return the contents
       unless log.nil? || !log.archived_at.nil?
         log_parts = Models::Log::Part.where(log_id: log.id)
-        log_parts.each { |log_part| contents << log_part.content }
-        contents
-      else
-        #go look in S3
+        log_data = []
+        log_parts.each { |log_part| log_data << log_part.content }
+        log_data
+      elsif log.archived_at?
+        # if it's not there then fetch it from S3.
+        archived_log_path = archive_url("/jobs/#{params[:job.id]}/log.txt")
+
+        content = open(Net::HTTP.get(URI.parse(archived_log_path)))
```

Comment:
This is good we should be able to use the S3 query object here pretty easily in the future (by passing it either the log object or the url) - I'm still working on it and should have it ready this week.